### PR TITLE
Improve docker local development

### PR DESCRIPTION
**Pull request by maestromac**
## What type of PR is this? (check all applicable)
- [x] Feature, DX

## Description
The work of this PR is heavily inspired by this blog post https://evilmartians.com/chronicles/ruby-on-whales-docker-for-ruby-rails-development

- Provides a 'bare-metal-like' development experience with a gem called `dip`.
- More intuitive setup by removing out-of-place services from docker-compose.
- Overall faster* (raw speed to be determined) to set up from the ground up.
- (Soon) adaptable to remote development environment.

## Related Tickets & Documents
n/a

## QA Instructions, Screenshots, Recordings
To test this locally, make sure you have docker for your OS installed, up and running.

```
# - make sure you disabled redis and postgresql so they won't fight over the ports
# - I also encourage backing up your current .env and start with fresh copy if there's issue spinning up the instance`
gem install dip
dip provision
dip up web

# try running a test
dip rspec spec/lib/acts_as_taggable_on

# check out other commands
dip ls
```

## Added tests?
- [x] no, because they aren't needed

## Added to documentation?
- [x] no documentation needed

## [optional] Are there any post deployment tasks we need to perform?
n/a

### CODE REVIEW

File: docker-compose.yml
Comment (maestromac) [inline docker-compose.yml:35]: Our repo defaults to `REDIS_URL="redis://localhost:6379" assuming the engineer is running redis locally. I can't seem to adjust this without it not working though. Is there an obvious answer I'm seeing?

File: dip.yml
Comment (maestromac) [inline dip.yml:75]: Temporarily comment this out because It's too much download 2 gig of dependency everything I trash the container.

File: docker-compose.yml
Comment (maestromac) [inline docker-compose.yml:10]: Future preping

File: Containerfile
Comment (klardotsh) [inline Containerfile:156]: TIL... neat.  Wild, but neat.

File: Containerfile
Comment (klardotsh) [inline Containerfile:174]: I don't like hard-coding `DISTRO_NAME` in as an `ARG` someone has to remember to update, I think, when `/etc/os-release` already has `VERSION_CODENAME=xxxxxx` we can grep+cut (or awk, or whatever) to pull out.  (I'd say to use the `lsb_release(1)` CLI but I get "no LSB modules are available" on Debian 12.1 when running that... ugh)

File: Containerfile
Comment (klardotsh) [inline Containerfile:188]: Why do we install vim here? Is this something that should be in a startup script (where the user can pick which editors they want)? Otherwise I fear this rapidly leads to "well why not neovim? why not emacs? why not helix?"

File: Containerfile
Comment (klardotsh) [inline Containerfile:181]: `PG_MAJOR` has no default value and could be unbound here, which would cause problems. I'd propose either providing a default at the image layer, or gracefully handling the empty case (perhaps by installing `postgresql-client`, no trailing dash, and letting Debian pick what the best stable version is)

File: docker-compose.yml
Comment (klardotsh) [inline docker-compose.yml:6]: Am I missing somewhere that this is used in the updated image? `RUBY_VERSION` is handled during base image builds for `ghcr.io/forem/ruby`, which happens in GHA CI, and should mostly never need re-plumbed here (I think).

File: docker-compose.yml
Comment (klardotsh) [inline docker-compose.yml:47]: IMO `vi` is not a user-friendly default here. Whether we use `nano`, [`micro`](https://micro-editor.github.io/), or something else entirely, is kinda up for debate, but `vi` is a brutal setting for `$EDITOR` - even vim/neovim would be a better bet. (I don't daily-drive Debian-based systems so it's possible that `vi` becomes a symlink to vim on Debian, but then let's not rely on it in `EDITOR`, and instead explicitly call vim... if we want vim as the git commit message editor and etc. all the other implications of setting `$EDITOR`)

**Review by klardotsh — COMMENTED** [review]
I'm not going to say I've run  this yet, but I think I might like to do so before a prospective merge. That said, I had enough commentary here that I think I'll hold off on doing that "integration test" to allow some conversation, first. Sorry it took so long to get you this review!

File: Containerfile
Comment (maestromac) [reply Containerfile:188]: Yeah this is completely optional. I will remove this
```suggestion
```

File: docker-compose.yml
Comment (maestromac) [reply docker-compose.yml:47]: Ok removing this default

File: docker-compose.yml
Comment (maestromac) [reply docker-compose.yml]: I overlooked this. You are right, I'm removing this from docker-compose.

File: Containerfile
Comment (maestromac) [reply Containerfile]: updated!

File: Containerfile
Comment (maestromac) [reply Containerfile:173]: changed it to use `$(lsb_release -cs)` instead, which is what's on PG's official doc https://www.postgresql.org/download/linux/debian/

Comment (klardotsh) [reply Containerfile]: Huh, I'm glad that works in the Docker container then. It didn't work in my QEMU VM :shrug:

Comment (maestromac) [reply Containerfile]: it doen't work on my fedora work station either :shrug:

**Review by klardotsh — COMMENTED** [review]
Gonna give this an actual run today and report back but I think this is the last of my "in theory" remarks.

File: .dockerdev/.bashrc
Comment (maestromac) [reply .dockerdev/.bashrc]: Removed

**Comment by klardotsh** [comment]
There's a weird glitch state `dip` gets into on my Linux (x86_64) system that seems to be related to the fact that I already have BuildKit running and working (since I work extensively with Docker already) and `dip`  tries to... start it back up I think? Not sure if maybe you've seen this error state already @maestromac. I was able to blindly retry and get around this, but it's a scary screen that's unpleasant UX and would scare off inbound contributors I think.

```
(woods) _forem1  » mac/update-containers » gem install --user dip
Successfully installed thor-1.2.2
Successfully installed dip-7.6.0
Parsing documentation for thor-1.2.2
Installing ri documentation for thor-1.2.2
Parsing documentation for dip-7.6.0
Installing ri documentation for dip-7.6.0
Done installing documentation for thor, dip after 0 seconds
2 gems installed

A new release of RubyGems is available: 3.4.10 → 3.4.19!
Run `gem update --system 3.4.19` to update your installation.

(woods) _forem1  » mac/update-containers » dip build
[+] Building 15.6s (3/3) FINISHED                                                                                                                                                                       docker-container:sharp_fermat
 => ERROR [sidekiq internal] booting buildkit                                                                                                                                                                                   15.6s
 => => pulling image moby/buildkit:buildx-stable-1                                                                                                                                                                              15.5s
 => => creating container buildx_buildkit_sharp_fermat0                                                                                                                                                                          0.0s
 => ERROR [rails internal] booting buildkit                                                                                                                                                                                     15.6s
 => => pulling image moby/buildkit:buildx-stable-1                                                                                                                                                                              15.5s
 => => creating container buildx_buildkit_sharp_fermat0                                                                                                                                                                          0.0s
 => CANCELED [webpacker internal] booting buildkit                                                                                                                                                                              15.6s
 => => pulling image moby/buildkit:buildx-stable-1                                                                                                                                                                              15.6s
 => => creating container buildx_buildkit_sharp_fermat0                                                                                                                                                                          0.0s
------
 > [sidekiq internal] booting buildkit:
------
------
 > [rails internal] booting buildkit:
------
Error response from daemon: Conflict. The container name "/buildx_buildkit_sharp_fermat0" is already in use by container "eea6e6346f9f3cd13d40f79d5adee6e40ecd3fece13742c08eb6040145a4b479". You have to remove (or rename) that container to be able to reuse that name.
                                                                                                                                                                                                                                      
(woods) _forem1  » mac/update-containers » dip build
[+] Building 121.4s (12/20)                                                                                                                                                                             docker-container:sharp_fermat
 => [sidekiq internal] booting buildkit                                                                                                                                                                                          1.8s
 => => starting container buildx_buildkit_sharp_fermat0                                                                                                                                                                          1.8s
 => [rails internal] booting buildkit                                                                                                                                                                                            1.8s
 => => starting container buildx_buildkit_sharp_fermat0                                                                                                                                                                          1.8s
 => [webpacker internal] booting buildkit                                                                                                                                                                                        1.8s
 => => starting container buildx_buildkit_sharp_fermat0                                                                                                                                                                          1.8s
 => [webpacker internal] load build definition from Dockerfile                                                                                                                                                                   0.0s
 => => transferring dockerfile: 8.57kB                                                                                                                                                                                           0.0s
 => [rails internal] load build definition from Dockerfile                                                                                                                                                                       0.0s

... onwards from here everything seems to work fine so far
```

**Review by klardotsh — APPROVED** [review]
Works. Didn't actually need the `up` incantation, `provision` is sufficient.

```
(woods) _forem1  » mac/update-containers » dip rspec spec/lib/acts_as_taggable_on
[+] Creating 2/0
 ✔ Container forem-redis-1     Running                                                                                                                                                     0.0s 
 ✔ Container forem-postgres-1  Running                                                                                                                                                     0.0s 
[Zonebie] Setting timezone: ZONEBIE_TZ="Warsaw"
......

Finished in 0.49101 seconds (files took 3.1 seconds to load)
6 examples, 0 failures

[TEST PROF INFO] Time spent in factories: 00:00.083 (11.6% of total time)
```

**Comment by maestromac** [comment]
@klardotsh Hmm I haven't seen that error before and this is the first time I see anything related buildkit outputted. I think this might be docker CE hiding those magics away.

**Comment by maestromac** [comment]
Admin merging this as it's development focus and there's more to come

**Pull request merged**
